sys/mathoid.js

```diff
@@ -17,37 +17,89 @@ function MathoidService(options) {
 }
 
 
+MathoidService.prototype._invalidateCache = function(hyper, hash) {
+
+    var routes = [];
+    var uri = '//wikimedia.org/api/rest_v1/media/math/';
+
+    routes.push(uri + 'formula/' + hash);
+
+    FORMATS.forEach(function(fmt) {
+        routes.push(uri + 'render/' + fmt + '/' + hash);
+    });
+
+    return hyper.post({
+        uri: new URI(['wikimedia.org', 'sys', 'events', '']),
+        body: routes.map(function(route) {
+            return { meta: { uri: route } };
+        })
+    }).catch(function(e) {
+        hyper.log('warn/bg-updates', e);
+    });
+
+};
+
+MathoidService.prototype.getFormula = function(hyper, req) {
+
+    var rp = req.params;
+    var hash = rp.hash;
+
+    return hyper.get({
+        uri: new URI([rp.domain, 'sys', 'post_data', 'mathoid.input', hash])
+    }).then(function(res) {
+        res.headers['x-resource-location'] = hash;
+        return res;
+    }).catch({ status: 404 }, function(err) {
+        // let's try to find an indirection
+        return hyper.get({
+            uri: new URI([rp.domain, 'sys', 'key_value', 'mathoid.hash_table', hash])
+        }).then(function(hashRes) {
+            // we have a normalised version of the formula
+            hash = hashRes.body;
+            // grab that version from storage
+            return hyper.get({
+                uri: new URI([rp.domain, 'sys', 'post_data', 'mathoid.input', hash])
+            }).then(function(res) {
+                res.headers['x-resource-location'] = hash;
+                return res;
+            });
+        });
+    });
+
+};
+
+
 MathoidService.prototype.checkInput = function(hyper, req) {
 
     var self = this;
     var rp = req.params;
-    var hash;
-    var origHash;
+    var hash = {};
+    var origHash = {};
     var checkRes;
 
     // start by calculating the hash
     return hyper.post({
         uri: new URI([rp.domain, 'sys', 'post_data', 'mathoid.input', 'hash']),
         body: { q: req.body.q, type: rp.type }
     }).then(function(res) {
-        hash = origHash = res.body;
+        hash.id = origHash.id = res.body;
```

Comment:
What's the point in making it an object? I don't see where any other properties are added or any difference from what happened.